### ISSUE TITLE
use diagnostics and localization

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8824]
Migrated from [rt.perl.org#41717](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=41717) (status was 'resolved')

Searchable as RT41717$

**Comment by p5pRT** [comment]
### From perl-5.8.0@ton.iguana.be

#### Created by perl-5.8.0@ton.iguana.be




perl \-le '$\_="abc"; while\(/\./g\) \{ print$a\,$& \}'
a
b
c

perl \-le 'use diagnostics; $\_="abc"; while\(/\./g\) \{ print$a\,$& \}'
a
a
a
\.\.\.\.

oops\, infinite loop
The reason is that use diagnostics localizes $\_ for the diagnostics output\,
but that looses pos on $\_

Easy enough to fix by going to diagnostics\.pm and in sub splainthis
replace the 

local $\_ = shift 

by a better localization\, e\.g\.

for \(shift\) \{
&nbsp;   \.\.\. the rest of splainthis comes here\.\.\.
\}

\(should work better in case of ties too\, irrespective of the pos issue\)

But\.\.\.shouldn't localization of $\_ just localize the pos too ?

Somewhat simular&#8203;:

perl \-le '$\!=21; 1\+$a;  print $\!' 2>/dev/full
Is a directory

perl \-le 'use diagnostics; $\!=21; 1\+$a;  print $\!' 2>/dev/full
No space left on device

So use diagnostics loses $\! too\. I think sub splainthis should also get a 
&nbsp;  local $\!

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

This perlbug was built using Perl v5.9.3 - Sun Jan 15 17:58:22 CET 2006
It is being executed now by  Perl 5.8.4 - Thu Jun  3 13:28:19 CEST 2004.

Site configuration information for perl 5.8.4:

Configured by ton at Thu Jun  3 13:28:19 CEST 2004.

Summary of my perl5 (revision 5 version 8 subversion 4) configuration:
  Platform:
    osname=linux, osvers=2.6.5, archname=i686-linux-64int-ld
    uname='linux quasar 2.6.5 #8 mon apr 5 05:41:20 cest 2004 i686 gnulinux '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=define use64bitall=undef uselongdouble=define
    usemymalloc=y, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -fomit-frame-pointer',
    cppflags='-fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='3.4.0 20031231 (experimental)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long long', ivsize=8, nvtype='long double', nvsize=12, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.3.2.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.3.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl 5.8.4:
    /usr/lib/perl5/5.8.4/i686-linux-64int-ld
    /usr/lib/perl5/5.8.4
    /usr/lib/perl5/site_perl/5.8.4/i686-linux-64int-ld
    /usr/lib/perl5/site_perl/5.8.4
    /usr/lib/perl5/site_perl
    .


Environment for perl 5.8.4:
    HOME=/home/ton
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/ton/bin.Linux:/home/ton/bin:/home/ton/bin.SampleSetup:/usr/local/bin:/usr/local/sbin:/home/oracle/product/9.2/bin:/usr/local/ar/bin:/usr/games/bin:/usr/X11R6/bin:/usr/share/bin:/usr/bin:/usr/sbin:/bin:/sbin:.
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @rgarcia

On 06/03/07\, via RT perl\-5\. 8\. 0 @&#8203; ton\. iguana\. be
\<perlbug\-followup@&#8203;perl\.org> >
> The reason is that use diagnostics localizes $\_ for the diagnostics output\,
> but that looses pos on $\_
>
> Easy enough to fix by going to diagnostics\.pm and in sub splainthis
> replace the
>
> local $\_ = shift
>
> by a better localization\, e\.g\.

Done as \#30491 \(using my\)\.

> \(should work better in case of ties too\, irrespective of the pos issue\)
>
> But\.\.\.shouldn't localization of $\_ just localize the pos too ?

Known bug\, won't fix\. In an ideal world you'd have 2 kinds of magic\,
the one that is localizable\, and the one that isn't\.\.\.

> Somewhat simular&#8203;:
>
> perl \-le '$\!=21; 1\+$a;  print $\!' 2>/dev/full
> Is a directory
>
> perl \-le 'use diagnostics; $\!=21; 1\+$a;  print $\!' 2>/dev/full
> No space left on device
>
> So use diagnostics loses $\! too\. I think sub splainthis should also get a
>   local $\!

Yes\, added as \#30492\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @nwc10

On Wed\, Mar 07\, 2007 at 12&#8203;:01&#8203;:44PM \+0100\, Rafael Garcia\-Suarez wrote&#8203;:
> On 06/03/07\, via RT perl\-5\. 8\. 0 @&#8203; ton\. iguana\. be

> >But\.\.\.shouldn't localization of $\_ just localize the pos too ?
> 
> Known bug\, won't fix\. In an ideal world you'd have 2 kinds of magic\,
> the one that is localizable\, and the one that isn't\.\.\.

Why did we say that we won't fix this? I don't remember the discussion\.
I take it that there's a non\-obvious good reason\, but I can't figure out
what\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @iabyn

On Wed\, Mar 07\, 2007 at 11&#8203;:16&#8203;:25AM \+0000\, Nicholas Clark wrote&#8203;:
> On Wed\, Mar 07\, 2007 at 12&#8203;:01&#8203;:44PM \+0100\, Rafael Garcia\-Suarez wrote&#8203;:
> > On 06/03/07\, via RT perl\-5\. 8\. 0 @&#8203; ton\. iguana\. be
> 
> > >But\.\.\.shouldn't localization of $\_ just localize the pos too ?
> > 
> > Known bug\, won't fix\. In an ideal world you'd have 2 kinds of magic\,
> > the one that is localizable\, and the one that isn't\.\.\.
> 
> Why did we say that we won't fix this? I don't remember the discussion\.
> I take it that there's a non\-obvious good reason\, but I can't figure out
> what\.

Some while ago I \(mostly\) fixed localisation of magic\, by distinguishing
between value and container magic\. One of the things that didn't get fixed
was pos magic \- it was more complex\, but I can't remember why now\.

looking at my List of Things to Do\, I see that pos magic is already on
it\, so perhaps it is fixable\.

\-\- 
Never do today what you can put off till tomorrow\.

**Comment by p5pRT** [comment]
### From @iabyn

On Wed\, Mar 07\, 2007 at 11&#8203;:24&#8203;:43AM \+0000\, Dave Mitchell wrote&#8203;:
> On Wed\, Mar 07\, 2007 at 11&#8203;:16&#8203;:25AM \+0000\, Nicholas Clark wrote&#8203;:
> > On Wed\, Mar 07\, 2007 at 12&#8203;:01&#8203;:44PM \+0100\, Rafael Garcia\-Suarez wrote&#8203;:
> > > On 06/03/07\, via RT perl\-5\. 8\. 0 @&#8203; ton\. iguana\. be
> > 
> > > >But\.\.\.shouldn't localization of $\_ just localize the pos too ?
> > > 
> > > Known bug\, won't fix\. In an ideal world you'd have 2 kinds of magic\,
> > > the one that is localizable\, and the one that isn't\.\.\.
> > 
> > Why did we say that we won't fix this? I don't remember the discussion\.
> > I take it that there's a non\-obvious good reason\, but I can't figure out
> > what\.
> 
> Some while ago I \(mostly\) fixed localisation of magic\, by distinguishing
> between value and container magic\. One of the things that didn't get fixed
> was pos magic \- it was more complex\, but I can't remember why now\.
> 
> looking at my List of Things to Do\, I see that pos magic is already on
> it\, so perhaps it is fixable\.

Fixed by changes \#30624\, \#30627\, \#30628

This code&#8203;:

&nbsp;    \{ local $x = 1; \.\.\. \}

now works roughly like the following&#8203;:

&nbsp;    \{
&nbsp;    my $saved\_ref = \\$x;
&nbsp;    \*x = \\1;
&nbsp;    if \($x was a container magic type\) \{
&nbsp;        attach a copy of $$saved\_ref's magic to $x;
&nbsp;        mg\_set\($x\);
&nbsp;    \}
&nbsp;    \.\.\.;

&nbsp;    \*x = $saved\_ref;
&nbsp;    if \($x was a container magic type\) \{
&nbsp;        mg\_set\($x\);
&nbsp;    \}
&nbsp;    \}

My fix was essentially to make that second mg\_set\(\) conditional\.


\-\- 
Britain\, Britain\, Britain\! Discovered by Sir Henry Britain in
sixteen\-oh\-ten\. Sold to Germany a year later for a pfennig and the promise
of a kiss\. Destroyed in eighteen thirty\-forty two\, and rebuilt a week
later by a man\. This we know\. Hello\. But what of the people of Britain?
Who they? What do? And why?   \-\- "Little Britain"
